### PR TITLE
feat: add support for Ray

### PR DESCRIPTION
**Pull request by zdtsw**
## Description

- move component name to api as const
- move create+reconcile compoent CR into func
-  cleanup old component files: ray and dashboard
- rename function to be more generic
- e2e test for ray: similar to dashboard
- update golangci for ireturn (if set inline get removed back and forth)

https://issues.redhat.com/browse/RHOAIENG-13182

## How Has This Been Tested?

local build: quay.io/wenzhou/opendatahub-operator-catalog:v2.16.13182-3

- install operator
- enable  dashbaord ray and kueue
```
   components:
    codeflare: {}
    kserve:
      serving:
        ingressGateway:
          certificate:
            type: OpenshiftDefaultIngress
        managementState: Managed
        name: knative-serving
    modelregistry:
      registriesNamespace: odh-model-registries
    trustyai: {}
    ray:
      managementState: Managed
    kueue:
      managementState: Managed
    workbenches: {}
    dashboard:
      managementState: Managed
    modelmeshserving: {}
    datasciencepipelines: {}
    trainingoperator: {}
```
- DSC status:
```
  installedComponents:
    dashboard: true
    ray: true
  phase: Ready
```
- 2 dashboard pods and 1 ray pod running in application namespace, no kueue pod
update ray CR set to
```
spec: {}
status:
  conditions:
    - lastTransitionTime: '2024-10-30T14:31:17Z'
      message: 1/1 deployments ready
      observedGeneration: 1
      reason: Ready
      status: 'True'
      type: Ready
  observedGeneration: 1
  phase: Ready

```

- update DSC to set Ray to Removed
- check ray CR again, it is gone
- check ray pod again, it is gone too
- check DSC
```
 installedComponents:
    dashboard: true
    ray: false
  phase: Ready

```

## Merge criteria

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [ ] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [ ] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: apis/components/v1/dashboard_types.go [hunk 1]
```diff
@@ -23,7 +23,8 @@ import (
 
 const (
 	// DashboardInstanceName the name of the Dashboard instance singleton.
-	// It must match what is set in the XValidation below.
+	DashboardComponentName = "dashboard"
```
Comment (lburgazzoli) [inline apis/components/v1/dashboard_types.go:26]: the comment above (about `DashboardInstanceName`) should probably moved down

File: pkg/controller/handlers/handlers.go [hunk 2]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/metadata/labels"
 )
 
-//nolint:ireturn
+
```
Comment (lburgazzoli) [inline pkg/controller/handlers/handlers.go:14]: this seems to have been removed by mistake

Comment (zdtsw) [reply pkg/controller/handlers/handlers.go:14]: yeah, it is back and forth, but i will fix it along with e2e tests

File: .golangci.yml [hunk 3]
```diff
@@ -58,6 +58,7 @@ linters-settings:
       - stdlib
       # also allow generics
       - generic
+      - EventHandler # for ToOwner
```
Comment (lburgazzoli) [inline .golangci.yml:61]: +1

File: controllers/datasciencecluster/datasciencecluster_controller.go [hunk 4]
```diff
@@ -308,22 +329,16 @@ func (r *DataScienceClusterReconciler) reconcileDashboardComponent(ctx context.C
...
+		r.Log.Error(err, "Failed to reconcile component: "+componentName)
 		instance = r.reportError(err, instance, fmt.Sprintf("failed to reconcile %s on DataScienceCluster", componentName))
 		instance, _ = status.UpdateWithRetry(ctx, r.Client, instance, func(saved *dscv1.DataScienceCluster) {
 			status.SetComponentCondition(&saved.Status.Conditions, componentName, status.ReconcileFailed, fmt.Sprintf("Component reconciliation failed: %v", err), corev1.ConditionFalse)
 		})
 		return instance, err
 	}
 
-	r.Log.Info("Dashboard component reconciled successfully")
+	r.Log.Info("component reconciled successfully: " + componentName)
```
Comment (lburgazzoli) [inline controllers/datasciencecluster/datasciencecluster_controller.go:341]: not for this PR, but we can probably squeeze the 3 `UpdateWithRetry` calls in a single one

**Comment by lburgazzoli** [comment]
LGTM

**Pull request merged**